src/struct_info.rs

```diff
@@ -265,64 +265,247 @@ impl<'a> StructInfo<'a> {
+                }});
+            let from_iter = from_iter.clone()
+            .into_configured(|span| {
+                syn::parse2(quote_spanned!(span.resolved_at(Span::mixed_site())=> |iter| iter.collect()))
+                    .expect("extend from_iter collect")
+            }).map(|Configured { keyword_span, value }| {
+                // A plain closure like `|iter| iter.collect()` would require a generic type annotation on the parameter type,
+                // which unfortunately isn't possible. This means the closure is repackaged as nested function.
+                let syn::ExprClosure {
+                    attrs,
+                    asyncness,
+                    movability,
+                    capture: _, // No locals are in scope here, so this doesn't matter.
+                    or1_token,
+                    inputs,
+                    or2_token,
+                    output,
+                    body,
+                } = value;
+
+                // A few error messages can be made much more direct on the way:
+
+                if let Some(asyncness) = asyncness {
+                    return Err(Error::new_spanned(asyncness, "Expected synchronous closure (Remove `async`)"))
+                }
 
-        let repeated_fields_error_type_name = syn::Ident::new(
-            &format!(
-                "{}_Error_Repeated_field_{}",
-                builder_name,
-                strip_raw_ident_prefix(field_name.to_string())
-            ),
-            proc_macro2::Span::call_site(),
-        );
-        let repeated_fields_error_message = format!("Repeated field {}", field_name);
+                if let Some(movability) = movability {
+                    return Err(Error::new_spanned(movability, "Expected movable closure (Remove `static`)"))
+                }
 
-        Ok(quote! {
-            #[allow(dead_code, non_camel_case_types, missing_docs)]
-            impl #impl_generics #builder_name < #( #ty_generics ),* > #where_clause {
-                #doc
-                pub fn #field_name (self, #field_name: #arg_type) -> #builder_name < #( #target_generics ),* > {
-                    let #field_name = (#arg_expr,);
-                    let ( #(#descructuring,)* ) = self.fields;
-                    #builder_name {
-                        fields: ( #(#reconstructing,)* ),
-                        _phantom: self._phantom,
+                let input = match inputs.len() {
+                    0 => return Err(Error::new_spanned(quote!(#or1_token #or2_token), "Expected one argument")),
+                    1 => inputs.into_iter().next().unwrap(),
+                    _ => return Err(Error::new_spanned(inputs, "Expected only one argument")),
+                };
+
+                if let syn::Pat::Type(syn::PatType { colon_token, ty, .. }) = input {
+                    return Err(Error::new_spanned(quote!(#colon_token #ty), "Did not expect argument type, since it is set explicitly by the macro"))
+                }
```

Comment:
I'll likely revisit this, making a closure argument type override the outer method argument type verbatim (including any `impl`).
That isn't quite standard Rust (though Syn should still parse it), but is needed to constrain the argument type for e.g. non-empty target collections.